File: Minimum Cost to Convert String I.py
# https://leetcode.com/problems/minimum-cost-to-convert-string-i/?envType=daily-question&envId=2024-07-27
# Minimum Cost to Convert String I

class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        
        if len(source) != len(target):
            return -1

        all_chars = set(source + target + ''.join(original) + ''.join(changed))

        inf = float('inf')
        min_cost = {char: {char2: inf for char2 in all_chars} for char in all_chars}

        for char in all_chars:
            min_cost[char][char] = 0

        for o, c, z in zip(original, changed, cost):
            min_cost[o][c] = min(min_cost[o][c], z)

        for k in all_chars:
            for i in all_chars:
                for j in all_chars:
                    if min_cost[i][j] > min_cost[i][k] + min_cost[k][j]:
                        min_cost[i][j] = min_cost[i][k] + min_cost[k][j]

        total_cost = 0

        for s_char, t_char in zip(source, target):
            if s_char == t_char:
                continue
            if min_cost[s_char][t_char] == inf:
                return -1
            total_cost += min_cost[s_char][t_char]

        return total_cost

File: Minimum Cost to Convert String I.js
// https://leetcode.com/problems/minimum-cost-to-convert-string-i/?envType=daily-question&envId=2026-01-29
// Minimum Cost to Convert String I

/**
 * @param {string} source
 * @param {string} target
 * @param {character[]} original
 * @param {character[]} changed
 * @param {number[]} cost
 * @return {number}
 */
var minimumCost = function(source, target, original, changed, cost) {
    const ALPHABET_SIZE = 26;
    const INF = Infinity;
    
    const dist = Array.from({ length: ALPHABET_SIZE }, () => 
        new Array(ALPHABET_SIZE).fill(INF)
    );
    
    for (let i = 0; i < ALPHABET_SIZE; i++) {
        dist[i][i] = 0;
    }
    
    for (let i = 0; i < original.length; i++) {
        const from = original[i].charCodeAt(0) - 97; // 'a'는 97
        const to = changed[i].charCodeAt(0) - 97;
        
        dist[from][to] = Math.min(dist[from][to], cost[i]);
    }
    
    for (let k = 0; k < ALPHABET_SIZE; k++) {
        for (let i = 0; i < ALPHABET_SIZE; i++) {
            for (let j = 0; j < ALPHABET_SIZE; j++) {
                if (dist[i][k] !== INF && dist[k][j] !== INF) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }
    
    let totalCost = 0;
    
    for (let i = 0; i < source.length; i++) {
        const u = source.charCodeAt(i) - 97;
        const v = target.charCodeAt(i) - 97;
        
        if (dist[u][v] === INF) {
            return -1;
        }
        
        totalCost += dist[u][v];
    }
    
    return totalCost;
};